TDataFrame.hxx

```diff
@@ -22,828 +22,1042 @@
+   TActionResultPtr<Double_t> Max(const std::string &branchName = "")
+   {
+      auto theBranchName(branchName);
       GetDefaultBranchName(theBranchName, "calculate the maximum");
       auto maxV = std::make_shared<T>(std::numeric_limits<T>::min());
-      return CreateAction<T, EActionType::kMax>(theBranchName,maxV);
+      return CreateAction<T, Internal::EActionType::kMax>(theBranchName, maxV);
    }
 
-   template<typename T = double>
-   TActionResultPtr<double> Mean(const std::string& branchName = "") {
-      auto theBranchName (branchName);
+   template <typename T = Double_t>
+   TActionResultPtr<Double_t> Mean(const std::string &branchName = "")
+   {
+      auto theBranchName(branchName);
       GetDefaultBranchName(theBranchName, "calculate the mean");
       auto meanV = std::make_shared<T>(0);
-      return CreateAction<T, EActionType::kMean>(theBranchName,meanV);
+      return CreateAction<T, Internal::EActionType::kMean>(theBranchName, meanV);
    }
 
 private:
+   /// Get the TDataFrameImpl if reachable. If not, throw.
+   std::weak_ptr<Details::TDataFrameImpl> GetDataFrameChecked()
+   {
+      auto df = fProxiedPtr->GetDataFrame();
+      if (df.expired()) {
+         throw std::runtime_error("The main TDataFrame is not reachable: did it go out of scope?");
+      }
+      return df;
+   }
 
-   void GetDefaultBranchName(std::string& theBranchName, const std::string& actionNameForErr) {
+   void GetDefaultBranchName(std::string &theBranchName, const std::string &actionNameForErr)
+   {
       if (theBranchName.empty()) {
          // Try the default branch if possible
-         const BranchList& defBl = fDerivedPtr->GetDataFrame().GetDefaultBranches();
+         const BranchList &defBl = fProxiedPtr->GetDataFrame().lock()->GetDefaultBranches();
          if (defBl.size() == 1) {
             theBranchName = defBl[0];
          } else {
-            std::string msg ("No branch in input to ");
+            std::string msg("No branch in input to ");
             msg += actionNameForErr;
-            msg +=  " and more than one default branch.";
+            msg += " and more than one default branch.";
             throw std::runtime_error(msg);
          }
       }
    }
 
-   template<typename BranchType, typename ActionResultType, enum EActionType, typename ThisType>
-   struct SimpleAction{ };
-
-   template<typename BranchType, typename ThisType>
-   struct SimpleAction<BranchType, TH1F, EActionType::kHisto1D, ThisType> {
-      static TActionResultPtr<TH1F> BuildAndBook(ThisType thisFrame, const std::string& theBranchName, std::shared_ptr<TH1F> h, unsigned nSlots) {
-         auto fillOp = std::make_shared<Operations::FillOperation>(h.get(), nSlots);
-         auto fillLambda = [fillOp](unsigned slot, const BranchType& v) mutable { fillOp->Exec(v, slot); };
+   template <typename BranchType, typename ActionResultType, enum Internal::EActionType, typename ThisType>
+   struct SimpleAction {};
```

Comment:
I guess this should be `*T*SimpleAction`. But probably we can get rid of `SimpleAction` altogether and have `BuildAndBook` be a member function of `TDataFrameInterface`, removing the need of passing `ThisType` and the `this` pointer around and simplifying the code a bit.

The trick is making `Internal::EActionType` a namespace with different structs named like the enum values, and using function overloading instead of partial specialization. [Working snippet](https://github.com/bluehood/tdataframe/files/708142/buildAndBook.txt).